src/type-data-frame.c

```diff
@@ -59,21 +59,48 @@ SEXP vctrs_new_data_frame(SEXP args) {
     Rf_errorcall(R_NilValue, "`x` must be a list");
   }
 
-  R_len_t size = df_size_from_list(x, n);
+  bool has_n = !Rf_isNull(n);
```

Comment:
We use `n == R_NilValue`

Reply:
Except where we don't ;-)

https://github.com/r-lib/vctrs/blob/9ac7d408266e7cfcd24d3ab995497856724adcfb/src/utils.c#L204

Will send a separate PR.

Reply:
No need for a PR.